Docs/pg stat progress copy (#14088)
yugabyte/yugabyte-db: docs/content/preview/explore/query-1-performance/pg-stat-progress-copy.md

---
title: View COPY status with pg_stat_progress_copy
linkTitle: View COPY status
description: Use pg_stat_progress_copy to get the COPY command status, number of tuples processed, and other COPY progress reports.
headerTitle: View COPY status with pg_stat_progress_copy
image: /images/section_icons/index/develop.png
menu:
  preview:
    identifier: pg-stat-progress-copy
    parent: query-tuning
    weight: 400
type: docs
---

> 

Reviewer: Maybe a link to COPY would useful

> Yugabyte supports the PostgreSQL `pg_stat_progress_copy` view to report the progress of the COPY command execution. Whenever COPY is running, the `pg_stat_progress_copy` view will contain one row for each backend that is currently running a COPY command.

Author: a client connection. Changed it.

For more information refer to [ROWS_PER_TRANSACTION](../../../api/ysql/the-sql-language/statements/cmd_copy/#rows-per-transaction).

### New column `yb_status`

The `pg_stat_progress_copy` view includes a status column `yb_status` to indicate the status of the copy command.

If a `COPY` command is terminated due to any error, then it is possible that some tuples are persisted as explained in the [`tuples_processed`](#definition-of-tuples-processed) section. In such scenarios, `tuples_processed` shows a non-zero count and `yb_status` will show that the `COPY` command was terminated due to an error. In PostgreSQL, it is not required as copy is one single transaction. For YugabyteDB, `yb_status` column helps finding if the copy successfully completed or not.

Following are the possible values for this column:

- IN PROGRESS – Indicating that the `COPY` command is still running.
- ERROR – `COPY` command was terminated due to an error.
- SUCCESS – `COPY` command successfully completed.

### `COPY` command information

The `pg_stat_progress_copy` view retains copy command information after the copy operation has completed.

> In PostgreSQL, after the `COPY` command is finished then the row containing details of the copy command is _removed_ from the view. In YugabyteDB, information is _retained_ in the view after the copy has finished. Also, information is retained only for the last executed copy command on that connection.

Reviewer: ```suggestion
In PostgreSQL, when the `COPY` command finishes, the row containing details of the copy command is _removed_ from the view. In YugabyteDB, the information is _retained_ in the view after the copy has finished. The information is retained only for the last executed copy command on that connection.
```

This is required for YugabyteDB, because if the `COPY` command finishes due to an error, then you would like to know how many rows were reliably persisted to the disk.

## Examples

> The following examples demonstrates the possible stages (IN PROGRESS, ERROR, SUCCESS) for the copy operation.

Reviewer: ```suggestion
The following examples demonstrate the possible stages (IN PROGRESS, ERROR, SUCCESS) for the copy operation.
```

### Prerequisites

- Make sure you have YugabyteDB up and running. If you are new to YugabyteDB, follow the steps in [Quick start](../../../quick-start/) to have YugabyteDB up and running in minutes.

- Run the following script from your YugabyteDB installation directory to generate data to a file `test_data.csv`.

    ```sh
    #!/bin/bash
    rm -rf test_data.csv
    for i in {1..100000}
    do
       echo "$i, $i, $i" >> test_data.csv
    done
    ```

> - From your local YugabyteDB installation directory, connect to the [YSQL](../../../../admin/ysqlsh/) shell, and create a table using the following schema:

Reviewer: bad link

    ```sql
    create table test_copy ( h1 int, r1 int, v1 int, primary key (h1, r1));
    ```

### Run a successful COPY command

1. Copy `test_data.csv` using the following copy command:

    ```sql
    \copy test_copy from 'test_data.csv'  (DELIMITER ',');
    ```

    ```output
    COPY 100000
    ```

1. Verify `yb_status` using the view.

    ```sql
    select * from pg_stat_progress_copy ;
    ```

    You should see output similar to the following:

    ```output

>      pid  | datid | datname  | relid |  command  | type | yb_status | bytes_processed | bytes_total | tuples_processed | tuples_excluded

Reviewer: table columns don't align properly

    -------+-------+----------+-------+-----------+------+-----------+-----------------+-------------+------------------+-----------------
    74390 | 13288 | yugabyte | 16394 | COPY FROM | PIPE | SUCCESS   |         1966685 |           0 |           100000 |               0
    (1 row)
    ```

### Verify a COPY command status in progress

Copy `test_data.csv` using the following copy command and check the status in parallel from another terminal.

```sql
\copy test_copy from 'test_data.csv'  (DELIMITER ',');
```

```sql
select * from pg_stat_progress_copy ;
```

You should see output similar to the following:

```output
 pid  | datid | datname  | relid |  command  | type |  yb_status  | bytes_processed | bytes_total | tuples_processed | tuples_excluded
-------+-------+----------+-------+-----------+------+-------------+-----------------+-------------+------------------+-----------------
74390 | 13288 | yugabyte | 16404 | COPY FROM | PIPE | IN PROGRESS |          766682 |           0 |            40000 |               0
(1 row)
```

### Verify an interrupted COPY operation

1. Copy `test_data.csv` and interrupt it using the following commands:

>     ```sh

Reviewer: sql

    \copy test_copy from 'test_data.csv'  (DELIMITER ',');
    ```

    ```output

>     ^CCancel request sent

Reviewer: CCancel - maybe separate the ^C from 'Cancel request sent' for clarity. I assume it's Ctrl+C? maybe spell that out in the step

    ERROR:  canceling statement due to user request
    CONTEXT:  COPY test_copy, line 55184: "55184, 55184, 55184"
    ````

1. Verify `yb_status` using the view.

     ```sql
    select * from pg_stat_progress_copy ;
    ```

    ```output
      pid  | datid | datname  | relid |  command  | type | yb_status | bytes_processed | bytes_total | tuples_processed | tuples_excluded
    -------+-------+----------+-------+-----------+------+-----------+-----------------+-------------+------------------+-----------------
     74390 | 13288 | yugabyte | 16399 | COPY FROM | PIPE | ERROR     |          766682 |           0 |            40000 |               0
    (1 row)
    ```

>     The default for `ROWS_PER_TRANSACTION` is 20000 and the interruption occurred on row 55184 in the example. So the last 15184 rows were not persisted and the first 40000 rows were persisted.

Reviewer 2: May be update this one as following --- 

The default for `ROWS_PER_TRANSACTION` is 2000 and the interruption occurred on row 55184 in the example.  As a result of this  first two transaction  of 20K rows were persisted resulting in 40K rows reported by tuples_processed. The next  in-progress transaction  was interrupted and due to this reason remaining 15184 rows were not persisted .

Reviewer 2: default for ROWS_PER_TRANSACTION is 20k . I noted wrong in comment above as 2k.

>     The default for `ROWS_PER_TRANSACTION` is 20K and the interruption occurred on row 55184 in the example. As a result, the first two transactions of 20K rows were persisted resulting in 40K rows reported by `tuples_processed`. The next _in-progress_ transaction was interrupted, and due to this reason the remaining 15184 rows were not persisted.

Reviewer: ```suggestion
    The default for `ROWS_PER_TRANSACTION` is 20K and the interruption occurred on row 55184 in the example. As a result, the first two transactions of 20K rows were persisted resulting in 40K rows reported by `tuples_processed`. The next _in-progress_ transaction was interrupted, and for this reason the remaining 15184 rows were not persisted.
```

| datname | Name of the database to which this backend is connected. |
| relid | OID of the table on which the COPY command is executed. It is set to 0 if copying from a SELECT query.|
| command | The command that is running COPY FROM, or COPY TO. |
| type | The I/O type that the data is read from or written to: FILE, PROGRAM, PIPE (for COPY FROM STDIN and COPY TO STDOUT), or CALLBACK (used for example during the initial table synchronization in logical replication). |
| yb_status | Tracking status of the copy command. |
| bytes_processed | Number of bytes already processed by COPY command. |
| bytes_total | Size of the source file for COPY FROM command in bytes. It is set to 0 if not available. |
| tuples_processed | Number of tuples already processed by COPY command. |
| tuples_excluded | Number of tuples not processed because they were excluded by the WHERE clause of the COPY command. |

## YugabyteDB specific changes

The `pg_stat_progress_copy` view includes the following changes which are YugabyteDB specific:

### Definition of `tuples_processed`

The definition of `tuples_processed` column is different in YugabyteDB in comparison to PostgreSQL.

> For the `COPY` command in YugabyteDB, an option `ROWS_PER_TRANSACTION` is added which defines the transaction size to be used by the `COPY` command. For example, if the total tuples to be copied are 5000 and `ROWS_PER_TRANSACTION` is set to 1000, then the database will create 5 transactions and each transaction will insert 1000 rows. If there is an error during the execution of the copy command, then some tuples can be persisted based on the already completed transaction. This also implies that if the error occurs after inserting the 3500th row, then the first 3000 rows will still be persisted in the database.

Reviewer 2: It is better to add further description in https://docs.yugabyte.com/preview/api/ysql/the-sql-language/statements/cmd_copy/  for the `ROWS_PER_TRANSACTION`  and link the page here. 

Also mention the default for ROWS_PER_TRANSACTION as well. 

IN 2.14/2.15  default is 20K and older releases as default as 1000. 

Finally we can on this page we can mention the high level summary with example. Duplicating  some text is also fine.

Author: @sushantrmishra , Thanks for all your suggestions. I have addressed them and if could you please look at this one in particular? I have included the details on both the pages and have linked them as well.

- 1 to 1000 →  Transaction_1
- 1001 to 2000 → Transaction_2
- 2001 to 3000 → Transaction_3
- 3001 to 3500 → Error

First 3000 rows will be persisted to the table and `tuples_processed` will show 3000.

Because the `copy` command is divided into multiple transactions, `tuples_processed` tracks the rows for which the transaction has already completed.

### New column `yb_status`

The `pg_stat_progress_copy` view includes a status column `yb_status` to indicate the status of the copy command.

> If a `COPY` command is terminated due to any error, then it is possible that some tuples are persisted as explained in the [`tuples_processed`](#definition-of-tuples-processed) section. In such scenarios, `tuples_processed` shows a non-zero count and `yb_status` will show that the `COPY` command was terminated due to an error. In PostgreSQL, it is not required as copy is one single transaction. However, for YugabyteDB, without this column, you can't determine if the copy successfully completed or not.

Reviewer 2: `However, for YugabyteDB, without this column, you can't determine if the copy successfully completed or not.` 

Rewrite as follows -- 

For YugabyteDB  yb_status columns helps finding   if the copy successfully completed or not.

     pid  | datid | datname  | relid |  command  | type | yb_status | bytes_processed | bytes_total | tuples_processed | tuples_excluded
    -------+-------+----------+-------+-----------+------+-----------+-----------------+-------------+------------------+-----------------
    74390 | 13288 | yugabyte | 16394 | COPY FROM | PIPE | SUCCESS   |         1966685 |           0 |           100000 |               0
    (1 row)
    ```

### Verify a COPY command status in progress

Copy `test_data.csv` using the following copy command and check the status in parallel from another terminal.

```sql
\copy test_copy from 'test_data.csv'  (DELIMITER ',');
```

```sql
select * from pg_stat_progress_copy ;
```

You should see output similar to the following:

```output
 pid  | datid | datname  | relid |  command  | type |  yb_status  | bytes_processed | bytes_total | tuples_processed | tuples_excluded
-------+-------+----------+-------+-----------+------+-------------+-----------------+-------------+------------------+-----------------
74390 | 13288 | yugabyte | 16404 | COPY FROM | PIPE | IN PROGRESS |          766682 |           0 |            40000 |               0
(1 row)
```

> ### Verify an interrupted COPY operation

Reviewer 2: Lets add that the  default for ROWS_PER_TRANSACTION is 20k, and interruption happened ROW 55184 in the given example, so last 15184 rows were not persisted and first 40K rows were persisted.

> 1. Copy `test_data.csv` and interrupt it using the following commands:

Reviewer: ```suggestion
1. Copy `test_data.csv` using the following command, and then interrupt it using Ctrl+C:
```

    ```sql
    \copy test_copy from 'test_data.csv'  (DELIMITER ',');
    ```

    ```output

>     Ctrl+C Cancel request sent

Reviewer: ```suggestion
    ^CCancel request sent
```

>     Ctrl+C Cancel request sent

Reviewer: if ^C actually shows up on the command line as shown. ^ = ctrl right?

Author: Yes , so do you suggest we retain it as it is or specify Ctrl?
